autogpt/llm/chat.py

```diff
@@ -55,206 +65,163 @@ def generate_context(prompt, relevant_memory, full_message_history, model):
-                agent.cycle_count,
-                current_context,
-                CURRENT_CONTEXT_FILE_NAME,
-            )
-
-            # TODO: use a model defined elsewhere, so that model can contain
-            # temperature and other settings we care about
-            assistant_reply = create_chat_completion(
-                model=model,
-                messages=current_context,
-                max_tokens=tokens_remaining,
-            )
-
-            # Update full message history
-            full_message_history.append(create_chat_message("user", user_input))
-            full_message_history.append(
-                create_chat_message("assistant", assistant_reply)
-            )
-
-            return assistant_reply
-        except RateLimitError:
-            # TODO: When we switch to langchain, this is built in
-            logger.warn("Error: ", "API Rate Limit Reached. Waiting 10 seconds...")
-            time.sleep(10)
+    """
+    Interact with the OpenAI API, sending the prompt, user input,
+        message history, and permanent memory.
+
+    Args:
+        system_prompt (str): The prompt explaining the rules to the AI.
+        user_input (str): The input from the user.
+        token_limit (int): The maximum number of tokens allowed in the API call.
+
+    Returns:
+    str: The AI's response.
+    """
+    model = cfg.fast_llm_model  # TODO: Change model from hardcode to argument
+    # Reserve 1000 tokens for the response
+    logger.debug(f"Token limit: {token_limit}")
+    send_token_limit = token_limit - 1000
+
+    # if len(agent.history) == 0:
+    #     relevant_memory = ""
+    # else:
+    #     recent_history = agent.history[-5:]
+    #     shuffle(recent_history)
+    #     relevant_memories = agent.memory.get_relevant(
+    #         str(recent_history), 5
+    #     )
+    #     if relevant_memories:
+    #         shuffle(relevant_memories)
+    #     relevant_memory = str(relevant_memories)
+    # logger.debug(f"Memory Stats: {agent.memory.get_stats()}")
+    relevant_memory = []
+
+    (
+        next_message_to_add_index,
+        current_tokens_used,
+        insertion_index,
+        message_chain,
+    ) = generate_context(system_prompt, relevant_memory, agent.history, model)
+
+    # while current_tokens_used > 2500:
+    #     # remove memories until we are under 2500 tokens
+    #     relevant_memory = relevant_memory[:-1]
+    #     (
+    #         next_message_to_add_index,
+    #         current_tokens_used,
+    #         insertion_index,
+    #         current_context,
+    #     ) = generate_context(
+    #         prompt, relevant_memory, agent.history, model
+    #     )
+
+    # Account for user input (appended later)
+    user_input_msg = create_chat_message("user", user_input)
+    current_tokens_used += count_message_tokens([user_input_msg], model)
+
+    current_tokens_used += 500  # Account for memory (appended later) TODO: The final memory may be less than 500 tokens
```

Comment:
Those 500 tokens are reserved for `new_summary_message`, because the length of `new_summary_message` is not known until after filling some space with bits of history and trimming+summarizing the rest.